python/tests/pyspark/lib/assertions.py

```diff
@@ -0,0 +1,35 @@
+import re
+from pandas.testing import assert_frame_equal
+
+
+def _print_dfs(actual, expected):
+    print('actual.show(100):')
+    actual.show(100, False)
+    print('expected.show(100):')
+    expected.show(100, False)
+
+
+def assert_df(actual, expected, sort=True, **kwargs):
+    """
+    For more elaborate DFs comparision, e.g. ignoring sorting or less precise.
```

Comment:
assert_frame_equal seems to do the job: it compares floats with restricted precision instead of full. Or am I missing how allclose would handle it better / easier?

Reply:
So why the comment here says "less precise" if assert_frame_equal does that already? I thought this testing function tried to introduce less precise assertions. Is this testing meant to just print out columns more clearly? I've never found this to be an issue for assert_frame_equal. Do you have an  example of this new output vs. the output of assert_frame_equal ?

Reply:
I can remove "less precise". There's no context for that left here, but used to be vs. direct comparison like this, for example:

```python
        assert sorted(list(actual.na.fill('null').toLocalIterator())) == \
               sorted(list(expected.na.fill('null').toLocalIterator()))
```

^ This would be too precise and easily fail on floating point differences.

Reply:
> Is this testing meant to just print out columns more clearly?

Not only that. This also sorts the df's (default sort=True) before comparing with assert_frame_equal.

> Do you have an example of this new output vs. the output of assert_frame_equal ?

I modified the `test_map` here to test.

Output with assert_df:

```
actual.show(100):
+-------+---------+---------+
|key_col|extra_col|value_col|
+-------+---------+---------+
|a      |b        |2.0      |
+-------+---------+---------+

expected.show(100):
+-------+---------+---------+
|key_col|extra_col|value_col|
+-------+---------+---------+
|a      |b        |1.0      |
+-------+---------+---------+



Ran 1 test in 10.802s

FAILED (failures=1)

Failure
Traceback (most recent call last):
  File "/Users/JuhoAutio/ideaprojects/mleap/python/tests/pyspark/lib/assertions.py", line 24, in assert_df
    assert_frame_equal(actual.toPandas(), expected.toPandas(), **kwargs)
  File "/Users/JuhoAutio/ideaprojects/mleap/python/venv/lib/python3.7/site-packages/pandas/util/testing.py", line 1502, in assert_frame_equal
    obj='DataFrame.iloc[:, {idx}]'.format(idx=i))
  File "/Users/JuhoAutio/ideaprojects/mleap/python/venv/lib/python3.7/site-packages/pandas/util/testing.py", line 1336, in assert_series_equal
    obj='{obj}'.format(obj=obj))
  File "pandas/_libs/testing.pyx", line 66, in pandas._libs.testing.assert_almost_equal
  File "pandas/_libs/testing.pyx", line 180, in pandas._libs.testing.assert_almost_equal
  File "/Users/JuhoAutio/ideaprojects/mleap/python/venv/lib/python3.7/site-packages/pandas/util/testing.py", line 1094, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame.iloc[:, 2] are different

DataFrame.iloc[:, 2] values are different (100.0 %)
[left]:  [2.0]
[right]: [1.0]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/JuhoAutio/.pyenv/versions/3.7.2/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/JuhoAutio/.pyenv/versions/3.7.2/lib/python3.7/unittest/case.py", line 615, in run
    testMethod()
  File "/Users/JuhoAutio/ideaprojects/mleap/python/tests/pyspark/string_map_test.py", line 29, in test_map
    assert_df(expected, result)
  File "/Users/JuhoAutio/ideaprojects/mleap/python/tests/pyspark/lib/assertions.py", line 33, in assert_df
    raise AssertionError("failed assert on column '{}': {}".format(actual.columns[int(iloc)], e))
AssertionError: failed assert on column 'value_col': DataFrame.iloc[:, 2] are different

DataFrame.iloc[:, 2] values are different (100.0 %)
[left]:  [2.0]
[right]: [1.0]
```

Output with `assert_frame_equal(expected.toPandas(), result.toPandas())`:

```
Traceback (most recent call last):
  File "/Users/JuhoAutio/.pyenv/versions/3.7.2/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/JuhoAutio/.pyenv/versions/3.7.2/lib/python3.7/unittest/case.py", line 615, in run
    testMethod()
  File "/Users/JuhoAutio/ideaprojects/mleap/python/tests/pyspark/string_map_test.py", line 32, in test_map
    assert_frame_equal(expected.toPandas(), result.toPandas())
  File "/Users/JuhoAutio/ideaprojects/mleap/python/venv/lib/python3.7/site-packages/pandas/util/testing.py", line 1502, in assert_frame_equal
    obj='DataFrame.iloc[:, {idx}]'.format(idx=i))
  File "/Users/JuhoAutio/ideaprojects/mleap/python/venv/lib/python3.7/site-packages/pandas/util/testing.py", line 1336, in assert_series_equal
    obj='{obj}'.format(obj=obj))
  File "pandas/_libs/testing.pyx", line 66, in pandas._libs.testing.assert_almost_equal
  File "pandas/_libs/testing.pyx", line 180, in pandas._libs.testing.assert_almost_equal
  File "/Users/JuhoAutio/ideaprojects/mleap/python/venv/lib/python3.7/site-packages/pandas/util/testing.py", line 1094, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame.iloc[:, 2] are different

DataFrame.iloc[:, 2] values are different (100.0 %)
[left]:  [2.0]
[right]: [1.0]
```

Now, the actual difference is easy enough to spot when the df's are small (not too many columns or rows).

However, I must say that even this is quite exotic. What if I would have two columns with some numeric values that I want to compare? It wouldn't be _that_ easy to say based on the index number in `DataFrame.iloc[:, 2]` where the problem lies. It doesn't print the column name `value_col` at all, and it doesn't print the full dataframes either.

Maybe there is some better alternative out there so that there would be no need for any of this custom assertion code(?).